lib/ai/embedding.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ai
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/sashabaranov/go-openai"
+
+	embeddingpb "github.com/gravitational/teleport/api/gen/proto/go/teleport/embedding/v1"
+	"github.com/gravitational/teleport/api/types"
+)
+
+const (
+	maxOpenAIEmbeddingsPerRequest = 1000
+	EmbeddingPeriod               = 5 * time.Minute
+)
+
+// Sha256Hash is the hash of the embedded content. This hash allows to detect if
+// the embedding is still up-to-date or if the content changed and the resource
+// must be re-embedded.
+type Sha256Hash = [32]byte
```

Comment:
drive-by nitpick

```suggestion
type Sha256Hash = [sha256.Size]byte
```

from `crypto/sha256`